### DISCUSSION TITLE
SIEVE implementation that allows concurrent access with no lock contention

### DISCUSSION THREAD

**Discussion by sakno** [cacheMon/NSDI24-SIEVE#4] [Show and tell]
Hi @1a1a11a,

First of all, I would like to thank you for SIEVE algorithm. I found its description in [this](https://cachemon.github.io/SIEVE-website/blog/2023/12/17/sieve-is-simpler-than-lru/#appendix) blog post. At first look, it's really simple and understandable. However, the devil is in the details. Naive implementation provided there in Python cannot be used in situation where multiple threads or co-routines trying to access the cache concurrently.

For instance, cache hit code
```python
node = self.cache[x]
node.visited = True
```

introduces race condition with the hand cursor adjustment:
```python
while obj and obj.visited:
    obj.visited = False
    obj = obj.prev if obj.prev else self.tail

```

Read of `visited` flag and then setting it to `False` is not atomic operation. For the real world application (especially when translating to Java/C/C++/C#) the flag must be analyzed and updated by using memory barriers and/or Compare-and-Swap operations.

But that's not all. The hand moves over the doubly-linked list. AFAIK, there is no lock-free implementation of update/delete operations for the linked list. Thus, concurrent insertions to the cache cannot be done in parallel due to lock contention. Moreover, the thread/co-routine which is trying to read the cached record can have race condition with the eviction triggered by concurrent thread. As a result, the reader can have stale reads or, which is more ugly, obtain the resource which is in the process of cleanup. There are ways to avoid stale reads. For instance, reference counting or epoch-based reclamation can be utilized. However, it requires some modifications of the original algorithm.

I would like to share my approach to deal with all these things:

- To avoid contention between eviction algorithm and insertion threads I've divided these components to different parallel jobs. The main idea is to replace lock contention with asynchronous decision about eviction.
- To avoid stale reads, I've split the record eviction into two phases: mark and sweep. Very similar to GC phases in many modern runtimes. In some cases, the mark and sweep phase can be done at the same time if the job responsible for maintaining SIEVE algorithm can ensure that Sweep phase is safe for the particular record.

Every thread that wants to insert a new value communicates with SIEVE algorithm asynchronously. To do that, I've organized lock-free queue. The queue has multiple producers (insertion threads) and a single consumer (a job that maintains SIEVE algorithm).

Full implementation you can find [here](https://github.com/dotnet/dotNext/tree/develop/src/DotNext.Threading/Runtime/Caching) in C#. Of course, it's much more complex than the original proposal.

**Comment by 1a1a11a** [comment]
Awesome work! The python code is used for illustration, but rather than production-ready code. It is great to see other engineers like you solving the challenges in making SIEVE concurrent. Do you want to add the repo to the website?

**Reply by sakno** [reply]
Yep, why not
